pkg/api/interceptor.go

```diff
@@ -86,62 +105,86 @@ func (wa *WalletAuthorizer) requiresAuthorization(req interface{}) bool {
+		return "", status.Errorf(codes.Unauthenticated, "authorization token is not provided")
 	}
 
 	words := strings.SplitN(values[0], " ", 2)
 	if len(words) != 2 {
-		return status.Errorf(codes.Unauthenticated, "invalid authorization header")
+		return "", status.Errorf(codes.Unauthenticated, "invalid authorization header")
 	}
 	if scheme := strings.TrimSpace(words[0]); scheme != "Bearer" {
-		return status.Errorf(codes.Unauthenticated, "unrecognized authorization scheme %s", scheme)
+		return "", status.Errorf(codes.Unauthenticated, "unrecognized authorization scheme %s", scheme)
 	}
 	token, err := decodeToken(strings.TrimSpace(words[1]))
 	if err != nil {
-		return status.Errorf(codes.Unauthenticated, "extracting token: %s", err)
+		return "", status.Errorf(codes.Unauthenticated, "extracting token: %s", err)
 	}
 
 	wallet, err := validateToken(ctx, wa.Log, token, time.Now())
 	if err != nil {
-		return status.Errorf(codes.Unauthenticated, "validating token: %s", err)
+		return "", status.Errorf(codes.Unauthenticated, "validating token: %s", err)
 	}
+	return wallet, nil
+}
 
+func (wa *WalletAuthorizer) authorize(ctx context.Context, req interface{}, wallet types.WalletAddr) error {
 	if pub, isPublish := req.(*messagev1.PublishRequest); isPublish {
 		for _, env := range pub.Envelopes {
 			if !wa.privilegedAddresses[wallet] && !allowedToPublish(env.ContentTopic, wallet) {
 				return status.Errorf(codes.PermissionDenied, "publishing to restricted topic")
 			}
 		}
 	}
-
-	return wa.authorizeWallet(ctx, wallet)
-}
-
-func (wa *WalletAuthorizer) authorizeWallet(ctx context.Context, wallet types.WalletAddr) error {
-	// * for limit exhaustion return status.Errorf(codes.ResourceExhausted, ...)
-	// * for other authorization failure return status.Errorf(codes.PermissionDenied, ...)
-
-	var allowListed bool
 	if wa.AllowLists {
 		if wa.AllowLister.IsDenyListed(wallet.String()) {
 			wa.Log.Debug("wallet deny listed", logging.WalletAddress(wallet.String()))
 			return status.Errorf(codes.PermissionDenied, ErrDenyListed.Error())
 		}
-		allowListed = wa.AllowLister.IsAllowListed(wallet.String())
 	}
+	return nil
+}
 
-	if !wa.Ratelimits {
-		return nil
+func (wa *WalletAuthorizer) applyLimits(ctx context.Context, req interface{}, wallet types.WalletAddr) error {
+	// * for limit exhaustion return status.Errorf(codes.ResourceExhausted, ...)
+	// * for other authorization failure return status.Errorf(codes.PermissionDenied, ...)
+
+	ip := clientIPFromContext(ctx)
+	if len(ip) == 0 {
+		// requests without an IP address are bucketed together as "ip_unknown"
+		ip = "ip_unknown"
+	}
+
+	// with no wallet apply regular limits
+	var priority bool
+	if len(wallet) > 0 && wa.AllowLists {
+		priority = wa.AllowLister.IsAllowListed(wallet.String())
+	}
+
+	// default request cost is 1,
+	// note that nil req (subscription stream) has cost 1 as well
+	cost := 1
+	switch req := req.(type) {
+	case *messagev1.PublishRequest:
```

Comment:
It looks like this will run against `Query` requests as well as `Publish`. If that's the case, we probably want to make the cost much higher for publish. For example, the WAF rules offer 10X the capacity for querying. Maybe we increase the rate of tokens and make publish cost 10 per envelope?